apm-agent-attach/src/main/java/co/elastic/apm/attach/ElasticApmAttacher.java

```diff
@@ -138,19 +142,30 @@ private static File getAgentJarFile() {
                 if (agentJar == null) {
                     throw new IllegalStateException("Agent jar not found");
                 }
-                // don't delete on exit, because this the attaching application may terminate before the target application
-                File tempAgentJar = File.createTempFile("elastic-apm-agent", ".jar");
-                try (OutputStream out = new FileOutputStream(tempAgentJar)) {
-                    byte[] buffer = new byte[1024];
-                    int length;
-                    while ((length = agentJar.read(buffer)) != -1) {
-                        out.write(buffer, 0, length);
+                String hash = md5Hash(ElasticApmAttacher.class.getResourceAsStream("/elastic-apm-agent.jar"));
+                File tempAgentJar = new File(System.getProperty("java.io.tmpdir"), "elastic-apm-agent-" + hash + ".jar");
+                if (!tempAgentJar.exists()) {
+                    try (OutputStream out = new FileOutputStream(tempAgentJar)) {
+                        byte[] buffer = new byte[1024];
+                        for (int length; (length = agentJar.read(buffer)) != -1;) {
+                            out.write(buffer, 0, length);
+                        }
                     }
                 }
                 return tempAgentJar;
-            } catch (IOException e) {
+            } catch (NoSuchAlgorithmException | IOException e) {
                 throw new IllegalStateException(e);
             }
         }
+
+    }
+
+    static String md5Hash(InputStream resourceAsStream) throws IOException, NoSuchAlgorithmException {
+        try (InputStream agentJar = resourceAsStream) {
+            MessageDigest md = MessageDigest.getInstance("MD5");
+            byte[] buffer = new byte[1024];
+            while (new DigestInputStream(agentJar, md).read(buffer) != -1) {}
```

Comment:
Can't the `DigestInputStream` be reused?

Reply:
Why would you want to reuse it?